cmd/mt-whisper-importer-reader/conversion.go

```diff
@@ -0,0 +1,209 @@
+	}
+
+	// Out of the input archives that we'll use, start with the lowest resolution one by converting
+	// it to the requested resolution and filling the resulting points into adjustedPoints.
+	// Then continue with archives of increasing resolutions while overwriting the generated points
+	// of previous ones.
+	for i := largestArchiveIdx; i >= smallestArchiveIdx; i-- {
+		in := c.points[i]
+		arch := c.archives[i]
+		if arch.SecondsPerPoint == spp {
+			if retIdx == 0 || c.method != "avg" {
+				for _, p := range in {
+					adjustedPoints[c.method][p.Timestamp] = p.Value
+				}
+			} else {
+				for _, p := range in {
+					adjustedPoints["sum"][p.Timestamp] = p.Value
+					adjustedPoints["cnt"][p.Timestamp] = 1
+				}
+			}
+		} else if arch.SecondsPerPoint > spp {
+			method := c.method
+			if method == "avg" && retIdx > 0 {
+				method = "fakeavg"
+			}
+			for m, points := range incResolution(in, method, arch.SecondsPerPoint, spp, rawRes) {
+				for _, p := range points {
+					adjustedPoints[m][p.Timestamp] = p.Value
+				}
+			}
+		} else {
+			methods := []string{c.method}
+			if c.method == "avg" && retIdx > 0 {
+				methods = []string{"sum", "cnt"}
+			}
+			for m, points := range decResolution(in, methods, arch.SecondsPerPoint, spp) {
+				for _, p := range points {
+					adjustedPoints[m][p.Timestamp] = p.Value
+				}
+			}
+		}
+
+	}
+
+	// merge the results that are keyed by timestamp into a slice of points
+	for m, p := range adjustedPoints {
+		for t, v := range p {
+			res[m] = append(res[m], whisper.Point{Timestamp: t, Value: v})
+		}
+		res[m] = sortPoints(res[m])
+	}
+
+	return res
+}
+
+// increase resolution of given points according to defined specs
+func incResolution(points []whisper.Point, method string, inRes, outRes, rawRes uint32) map[string][]whisper.Point {
+	out := make(map[string][]whisper.Point)
+	aggFactor := float64(outRes) / float64(rawRes)
+	for _, inPoint := range points {
+		if inPoint.Timestamp == 0 {
+			continue
+		}
+
+		rangeEnd := inPoint.Timestamp - (inPoint.Timestamp % outRes)
+
+		var outPoints []whisper.Point
+		for ts := rangeEnd; ts > inPoint.Timestamp-inRes; ts = ts - outRes {
+			outPoints = append(outPoints, whisper.Point{Timestamp: ts})
+		}
+
+		for _, outPoint := range outPoints {
+			if method == "cnt" || method == "sum" {
+				outPoint.Value = inPoint.Value / float64(len(outPoints))
+			} else if method == "fakeavg" {
+				outPoint.Value = inPoint.Value * aggFactor
+			} else {
+				outPoint.Value = inPoint.Value
+			}
+
+			if method == "fakeavg" {
+				out["sum"] = append(out["sum"], outPoint)
+				out["cnt"] = append(out["cnt"], whisper.Point{Timestamp: outPoint.Timestamp, Value: aggFactor})
+			} else {
+				out[method] = append(out[method], outPoint)
+			}
+		}
+	}
+	for m := range out {
+		out[m] = sortPoints(out[m])
+	}
+	return out
+}
+
+func decResolution(points []whisper.Point, methods []string, inRes, outRes uint32) map[string][]whisper.Point {
```

Comment:
can you document these functions. it's easy to confuse "decrease resolution" to "decrease interval", whereas it's the opposite: we're increasing the interval here.

sidenote:
https://www.google.be/search?q=resolution&oq=resolution&aqs=chrome..69i57j0l5.1000j0j7&sourceid=chrome&ie=UTF-8 has 1 applicable definition for resolution:

> the smallest interval measurable by a telescope or other scientific instrument; the resolving power.
the degree of detail visible in a photographic or television image.

this definition seems also inconsistent to me. as they use resolution for both "interval" and "degree of detail", but the lower the interval, the higher the degree of detail..

and we also speak of "high resolution screens" to talk about smaller intervals/pixels, so either way of phrasing seems correct, but just worth commenting what exactly the functions do.

Reply:
actually i think "decrease interval" is kind of ambiguous because strictly speaking it increases the resolution, as you say, but i think many people might misunderstand it as decreasing the resolution.

Reply:
since resolution is not well defined and ambiguous, i don't understand what you're saying.
but i have no trouble believing that people can misunderstand interval - even though it's more clearly defined than resolution, so that's why i'm saying to just document the functions better.

Reply:
added comments